Workaround Flux#1027

ericphanson opened the pull request:
This example very much does not work:
```julia
julia> output
10×1 Matrix{Float32}:
 1.2540959f-6
 5.1405354f-5
 0.00020041084
 3.131654f-5
 5.9973813f-6
 4.5418696f-7
 5.347429f-8
 0.99964094
 3.7541522f-5
 3.067953f-5

julia> output2
10×1 Matrix{Float32}:
 0.086956024
 0.10940457
 0.09150873
 0.098902285
 0.088835925
 0.06954686
 0.06539499
 0.20102207
 0.090939134
 0.0974893
```

The goal of this PR is to fix things so that we can correctly (de)-serialize this model so the test passes.

ericphanson commented:
This works around https://github.com/FluxML/Flux.jl/issues/1027 by essentially following the strategy of https://github.com/FluxML/Flux.jl/pull/1509#issuecomment-798947090, where here I've called @touchesir's `actually_all_params` just `weights` and added `loadweights!` as an analog to `loadparams!` that uses these.

This adds a Flux dependency here which is a little unfortunate because downstream code might want to use the LegolasFlux schema without actually touching weights/flux stuff (e.g. looking at a bunch of losses or something), and now such code will pull in the Flux dependency. But when upstream fixes #1027 we can probably remove the `flux_workarounds.jl` file (and make a breaking release).

ericphanson commented on Project.toml:
```diff
@@ -1,22 +1,27 @@
 uuid = "eb5f792d-d1b1-4535-bae3-d5649ec7daa4"
 authors = ["Beacon Biosignals, Inc."]
-version = "0.1.0"
+version = "0.1.1"
```
non-breaking; all existing usage of LegolasFlux will work, but if you switch to the new `weights` / `loadweights!` it'll even work correctly with `BatchNorm`s 😄

ericphanson commented on examples/digits.jl:
```diff
@@ -0,0 +1,120 @@
+N_train = 1_000
+N_test = 50
+
+##
+# to use MNIST data, uncomment these
+# train_x, train_y = MNIST.traindata(Float32, 1:N_train)
+# test_x,  test_y  = MNIST.testdata(Float32, 1:N_test)
```
decided to avoid downloading data in CI / any possible license issues with MNIST (MLDatasets gives you a warning that the license is ambiguous)

ToucheSir commented on src/flux_workarounds.jl:
```diff
@@ -0,0 +1,57 @@
+"""
+function other_weights end
+
+other_weights(layer) = ()
```
Re Flux dependency, the only dependency that is strictly necessary here is Functors.jl itself. `Functors.children(layer)` will return _everything_ that layer considers a parameter, so you shouldn't be missing out on anything :)

ToucheSir commented on src/flux_workarounds.jl:
````diff
@@ -0,0 +1,57 @@
+# ```
+# to `weights!(p::Params, x, seen = IdSet())`.
+
+weights!(p::Params, x::AbstractArray{<:Number}, seen = IdSet()) = push!(p, x)
````
Instead of pushing to `Params` (which requires a dep on Zygote), you can just push to a plain vector of arrays. I imagine that would also be much easier to serialize as well!

ToucheSir replied:
I just occurred to me that `weights!` is very similar to [`fcollect(...; exclude=Functors.isleaf)`](https://github.com/FluxML/Functors.jl/blob/master/src/functor.jl#L108). Would that work for your use-case?

ToucheSir reviewed (commented):
Thanks for the ping! I've left a couple of comments that might be helpful.

hannahilea commented on src/flux_workarounds.jl:
```diff
@@ -0,0 +1,57 @@
+Given a layer with params that are not captured by `Flux.trainable`, produce
+a tuple of arrays corresponding to these parameters (analogous to `Flux.trainable`).
+"""
+function other_weights end
```
what does this line do, given the default implementation below?

ericphanson replied:
I did this so the docstring would be attached to the function as a whole, instead of a particular method. It doesn't matter here since none of the others have docstrings anyway, but we could add a separate docstring for each method and if you do `? other_weights` or `? other_weights(layer)` it will show you all the docstrings for all the methods that fit that signature, so we could just put it on the fallback method and it would always show up, just like it will when we attach it to the function itself like I did here (see https://docs.julialang.org/en/v1/manual/documentation/#Functions-and-Methods for probably a better explanation). But really I just did it because I felt like conceptually that docstring belongs to the whole function and I felt like it helped me formulate the API. Like here's the function, here's what it does in words, and then here are the methods to do that.

hannahilea replied:
oh cool! thanks for explaining. :)

hannahilea commented on test/runtests.jl:
```diff
@@ -43,3 +45,37 @@ end
+            Flux.loadparams!(r2, p)
+            testmode!(r2)
+
+            @test_broken output ≈ r2(x)
```
Nice! Might be worth adding a comment here that iff this test _fails_, we can reevaluate use of flux workaround, and linking to the description you wrote in this issue?

ericphanson replied:
done!

hannahilea reviewed (changes requested):
Nice!

Other than the renaming issue below, this seems good to go. :)

ericphanson commented:
Thanks for the reviews!

Thanks especially for the pointers to `fcollect` and Functors, @ToucheSir! Super helpful. In the last two commits, I switched to using that approach and dropped the Flux dep (for the very light Functors dep). I found that `fcollect` as-is does not work because of this line: https://github.com/FluxML/Functors.jl/blob/adeb24bc3b2fb3e9959f1157d81f4633a855e207/src/functor.jl#L109. `x in arr` for arrays `arr` compares by value, and if we end up with several `x`s with the same values (say zero'd out arrays from a freshly initialized model or something like that), then we only keep 1 of each, and when it comes time to load the weights later, we don't have enough of them (this isn't hypothetical, it happened with the DigitsModel example here and had me confused for awhile). So instead in [`fcollect2`](https://github.com/beacon-biosignals/LegolasFlux.jl/pull/4/files#diff-1762840cbf23144855a52c56e365f6fc4358ea3d6307893157c2e6027aa73953R1-R11) I switched to using an `IdSet` to keep track of the objects we've already added (similarly to `fmap` above). Since an `IdSet` is unordered and we definitely care about the ordering, I had to also keep the original `Vector{Any}` used to collect the weights. Keeping them in two data structures is a little inelegant (I guess I actually want an `OrderedIdSet` which doesn't exist AFAIK) but should have a negligible perfomance cost in this context.

Also, since we now have `loadweights!` available in the package, we might be able to make the API a little easier to use (e.g. allow a user to pass a model directly instead of making them get the weights out first themselves). But I think that should be followup work (I think it requires a little more thought about the best approach).

ToucheSir commented:
Nice catch for `fcollect`. IMO this also a bug in the upstream implementation and we should fix it to work like `fcollect2`. Will have a look at that when I next find time to work on Functors.

ericphanson commented:
Ok good to hear this is likely a bug! I've filed an issue, https://github.com/FluxML/Functors.jl/issues/16.

ararslan commented on README.md:
```diff
@@ -28,16 +29,16 @@ my_model = make_my_model()
 # We can save whatever other columns we'd like to as well as the `weights`.
-model_row = ModelRow(; weights = collect(params(cpu(my_model))), architecture_version = 1, loss = 0.5)
+model_row = ModelRow(; weights = weights(cpu(my_model)),
+                       architecture_version = 1, loss = 0.5)
```
```suggestion
                     architecture_version=1, loss=0.5)
```

ararslan commented on examples/digits.jl:
```diff
@@ -0,0 +1,140 @@
+# Construct the actual model from a config object. This is the only
+# constructor that should be used, to ensure the model is created just
+# from the config object alone.
+function DigitsModel(config::DigitsConfig = DigitsConfig())
```
```suggestion
function DigitsModel(config::DigitsConfig=DigitsConfig())
```

ararslan commented on examples/digits.jl:
```diff
@@ -0,0 +1,140 @@
+function DigitsModel(config::DigitsConfig = DigitsConfig())
+    dropout_rate = config.dropout_rate
+    Random.seed!(config.seed)
+    chain = Chain(
+        Dropout(dropout_rate),
+        Conv((3, 3), 1=>32, relu),
+        BatchNorm(32, relu),
+        MaxPool((2,2)),
+        Dropout(dropout_rate),
+        Conv((3, 3), 32=>16, relu),
+        Dropout(dropout_rate),
+        MaxPool((2,2)),
+        Dropout(dropout_rate),
+        Conv((3, 3), 16=>10, relu),
+        Dropout(dropout_rate),
+        x -> reshape(x, :, size(x, 4)),
+        Dropout(dropout_rate),
+        Dense(90, 10), softmax)
```
```suggestion
    chain = Chain(Dropout(dropout_rate),
                  Conv((3, 3), 1 => 32, relu),
                  BatchNorm(32, relu),
                  MaxPool((2, 2)),
                  Dropout(dropout_rate),
                  Conv((3, 3), 32 => 16, relu),
                  Dropout(dropout_rate),
                  MaxPool((2, 2)),
                  Dropout(dropout_rate),
                  Conv((3, 3), 16 => 10, relu),
                  Dropout(dropout_rate),
                  x -> reshape(x, :, size(x, 4)),
                  Dropout(dropout_rate),
                  Dense(90, 10),
                  softmax)
```

ararslan commented on examples/digits.jl:
```diff
@@ -0,0 +1,140 @@
+    loss = (x, y) -> crossentropy(m(x), y)
+    opt = ADAM()
+    evalcb = throttle(() -> @show(accuracy(m, tX, tY)), 5)
+    Flux.@epochs 1 Flux.train!(loss, params(m), Iterators.take(train, N), opt, cb = evalcb)
```
```suggestion
    Flux.@epochs 1 Flux.train!(loss, params(m), Iterators.take(train, N), opt; cb=evalcb)
```

ararslan commented on examples/digits.jl:
```diff
@@ -0,0 +1,140 @@
+m = DigitsModel()
+
+# increase N to actually train more than a tiny amount
+acc = train_model!(m; N = 10)
```
```suggestion
acc = train_model!(m; N=10)
```

ararslan commented on src/functors.jl:
```diff
@@ -0,0 +1,39 @@
+# Modified version of `fcollect` to use an `IdSet` cache so that
+# distinct arrays whose values happen to be duplicates are each kept.
+# <https://github.com/FluxML/Functors.jl/issues/16>
+function fcollect2(x; output = [], cache = Base.IdSet(), exclude = v -> false)
```
```suggestion
function fcollect2(x; output=[], cache=IdSet(), exclude=_ -> false)
```

ararslan commented on src/functors.jl:
```diff
@@ -0,0 +1,39 @@
+function fcollect2(x; output = [], cache = Base.IdSet(), exclude = v -> false)
+    x in cache && return output
+    if !exclude(x)
+      push!(cache, x)
+      push!(output, x)
+      foreach(y -> fcollect2(y; cache = cache, output=output, exclude = exclude), Functors.children(x))
```
```suggestion
        push!(cache, x)
        push!(output, x)
        foreach(y -> fcollect2(y; cache=cache, output=output, exclude=exclude), Functors.children(x))
```

ararslan commented on src/functors.jl:
```diff
@@ -0,0 +1,39 @@
+function loadweights!(m, xs)
+    for (i, (p, x)) in enumerate(zip(weights(m), xs))
+        if size(p) != size(x)
+            error("For the $(i)th weight expected param size $(size(p)), got $(size(x))")
```
Perhaps throw an informative error type, e.g. `ArgumentError`?

ericphanson replied:
done!

ararslan commented on test/runtests.jl:
```diff
@@ -43,3 +46,42 @@ end
+        model = mk_model()
+        trainmode!(model)
+        x = reshape([1f0], 1, 1, 1)
+        for i = 1:10
```
```suggestion
        for i in 1:10
```

hannahilea commented on src/functors.jl:
```diff
@@ -0,0 +1,43 @@
+function load_weights!(m, xs)
+    model_weights = weights(m)
+    if length(model_weights) != length(xs)
+        throw(ArgumentError("Number of weights given ($(length(xs))) does not match number of weights model expects ($(length(model_weights)))"))
```
:)

hannahilea replied:
safety first! :)

ericphanson replied:
💯

ToucheSir replied:
Yes please, not doing this has lead to subtle errors in `loadparams!` in upstream Flux :)

hannahilea commented on test/runtests.jl:
```diff
@@ -3,6 +3,7 @@ using Test
 using Flux, LegolasFlux
 using LegolasFlux: Weights, FlatArray, ModelRow
 using Arrow
+using Random
```
Is it possible we want to use StableRNG here (and then seeded rands, below)?

ericphanson replied:
good idea, done!

hannahilea reviewed (changes requested):
...so I wrote out a whole big comment and was _sure_ it was attached to my previous review, but now I don't see it anywhere? The gist is that `weights` has a very specific meaning in the context of ML, so our use of `weights` (in `load_weights(...)`, `weights(...)`, etc) should probably actually be renamed to something more like `parameters` or `values`. (Too bad `params` is already taken by Flux!)

Basically, while all weights are parameters, not all parameters are weights (some are biases, etc).

ToucheSir commented:
> (Too bad `params` is already taken by Flux!)

Not only that, the name `params` is misleading because it only returns trainable parameters. Moreover, it conflicts with `BenchmarkTools.params` as well. I wish we could rename it to something else, but it would be a massive back compat break because `params` is arguably the most used function in all of Flux (not counting re-exports from Zygote etc.)

ericphanson commented:
Ok, we've got some bikeshedding to do! Some ideas...

|           | Getter            | Setter            | Column name |
|-----------|-------------------|-------------------|-------------|
| Current   | `weights`         | `load_weights!`   | `weights`   |
| state     | `fetch_state`     | `load_state!`     | `state`     |
| learnings | `fetch_learnings` | `load_learnings!` | `learnings` |

We're leaning towards `state` but if anyone has any further ideas or comments let me know! I will pick something and merge tomorrow morning if there aren't other objections.

hannahilea commented:
...pytorch uses `state`:
> In PyTorch, the learnable parameters (i.e. weights and biases) of an torch.nn.Module model are contained in the model’s parameters (accessed with model.parameters()). A state_dict is simply a Python dictionary object that maps each layer to its parameter tensor. Note that only layers with learnable parameters (convolutional layers, linear layers, etc.) and registered buffers (batchnorm’s running_mean) have entries in the model’s state_dict. Optimizer objects (torch.optim) also have a state_dict, which contains information about the optimizer’s state, as well as the hyperparameters used.
 (https://pytorch.org/tutorials/beginner/saving_loading_models.html#what-is-a-state-dict)

...tensorflow uses `weights` (huh):
> A Keras model consists of multiple components:
> 
> The architecture, or configuration, which specifies what layers the model contain, and how they're connected.
> A set of weights values (the "state of the model"). (https://www.tensorflow.org/guide/keras/save_and_serialize#introduction)

so maybe I sent us down this bikeshed prematurely, and it is okay to stick with `weights` after all... 🤷

hannahilea reviewed (approved):
...in light of the renaming discussion above, I'm going to leave the renaming (or not) to your discretion and let you merge at your leisure.

ericphanson commented:
Ok cool! Thanks for looking up how others do it, that was a great idea.

I think then since `weights` seems acceptable I'd like to go with that. We only allow (multi-dimensional) arrays with all the same element type, which is far from arbitrary state. That restriction is on purpose of course, since we are trying to provide a different way to serialize the model rather than include arbitrary state from your julia session, so I think it's good to highlight it a little. Also, it's nice to keep this non-breaking by not adjusting the schema.

But I think I'd like to use `fetch_weights` instead of `weights` so that we have `fetch_weights`, `load_weights!`, and column name `weights`. I like that better than reusing `weights` for both the function and the column name because you can then do
```julia
weights = fetch_weights(model)

row = ModelRow(; weights, ...)
```
